asdf/extension/_legacy.py

```diff
@@ -117,12 +117,17 @@ def __init__(self, extensions):
         for extension in extensions:
             tag_mapping.extend(extension.tag_mapping)
             url_mapping.extend(extension.url_mapping)
-            for typ in extension.types:
-                self._type_index.add_type(typ, extension)
-                validators.update(typ.validators)
-                for sibling in typ.versioned_siblings:
-                    self._type_index.add_type(sibling, extension)
-                    validators.update(sibling.validators)
+            if extension.legacy:
+                for typ in extension.types:
+                    self._type_index.add_type(typ, extension)
+                    validators.update(typ.validators)
+                    for sibling in typ.versioned_siblings:
+                        self._type_index.add_type(sibling, extension)
+                        validators.update(sibling.validators)
+            else:
+                # for non-legacy extensions, get validators from the extension
+                # not the types
+                validators.update(extension.validators)
```

Comment:
```suggestion
            if extension.legacy:
                for type_ in extension.types:
                    self._type_index.add_type(type_, extension)
                    validators.update(type_.validators)
                    for sibling in type_.versioned_siblings:
                        self._type_index.add_type(sibling, extension)
                        validators.update(sibling.validators)
            else:
                # for non-legacy extensions, get validators from the extension
                # not the types
                validators.update(extension.validators)
```

I find `typ` to be jarring when used in place of `type` (which we are avoiding due to collision with the builtin). I suggest using the standard work-around which is appending `_` to the end. ie `type_`.